File: scratch/sync_classes.py
import os
from supabase import create_client

url = os.environ.get("EXPO_PUBLIC_SUPABASE_URL")
key = os.environ.get("EXPO_PUBLIC_SUPABASE_KEY")
supabase = create_client(url, key)

def fix_classes():
    # 1. Get all students
    students = supabase.table('students').select('id, class_level').execute().data
    
    for student in students:
        s_id = student['id']
        c_level = student['class_level']
        
        # 2. Get distinct subjects for this class level from timetable_periods
        periods = supabase.table('timetable_periods').select('subject_name, teacher_name').eq('class_level', c_level).execute().data
        
        # Map of subject -> teacher
        subjects = {}
        for p in periods:
            subj = p['subject_name']
            if subj not in subjects:
                subjects[subj] = p['teacher_name'] or "Mr. Smith"
        
        # 3. Ensure a class exists for each subject + class_level
        class_ids = []
        for subj, teacher in subjects.items():
            class_name = f"Class {c_level} {subj}"
            
            # check if exists
            existing = supabase.table('classes').select('id').eq('name', class_name).execute().data
            if existing:
                class_ids.append(existing[0]['id'])
            else:
                # create
                new_class = supabase.table('classes').insert({
                    'name': class_name,
                    'subject': subj,
                    'teacher_name': teacher,
                    'room': f"Room {c_level}0{list(subjects.keys()).index(subj)+1}",
                    'schedule': "Mon-Fri"
                }).execute().data
                if new_class:
                    class_ids.append(new_class[0]['id'])
        
        # 4. Link to student_classes
        # first delete existing to avoid duplicates
        supabase.table('student_classes').delete().eq('student_id', s_id).execute()
        
        for cid in class_ids:
            supabase.table('student_classes').insert({
                'student_id': s_id,
                'class_id': cid
            }).execute()
            
    print("Done populating classes for all students based on their timetable.")

if __name__ == '__main__':
    fix_classes()

File: scratch/sync_classes.js
import { createClient } from '@supabase/supabase-js';
import dotenv from 'dotenv';

dotenv.config();

const supabase = createClient(process.env.EXPO_PUBLIC_SUPABASE_URL, process.env.EXPO_PUBLIC_SUPABASE_KEY);

async function fixClasses() {
  const { data: students, error: studentsErr } = await supabase.from('profiles').select('id, class_level').eq('role', 'student');
  if (studentsErr) throw studentsErr;

  for (const student of students) {
    const sId = student.id;
    const cLevel = student.class_level;
    
    if (!cLevel) continue;

    const { data: periods, error: periodsErr } = await supabase.from('timetable_periods').select('subject_name, teacher_name').eq('class_level', cLevel);
    if (periodsErr) throw periodsErr;

    const subjects = {};
    for (const p of periods) {
      if (!subjects[p.subject_name]) {
        subjects[p.subject_name] = p.teacher_name || "Mr. Smith";
      }
    }

    const classIds = [];
    let idx = 1;
    for (const [subj, teacher] of Object.entries(subjects)) {
      const className = `Class ${cLevel} ${subj}`;

      const { data: existing, error: existErr } = await supabase.from('classes').select('id').eq('name', className);
      if (existErr) console.error("Exist err:", existErr);
      
      if (existing && existing.length > 0) {
        classIds.push(existing[0].id);
      } else {
        const { data: newClass, error: insErr } = await supabase.from('classes').insert({
          name: className,
          subject: subj,
          teacher_name: teacher,
          room: `Room ${cLevel}0${idx}`,
          schedule: "Mon-Fri"
        }).select();
        
        if (insErr) {
            console.error("Insert class err:", insErr);
        }
        
        if (newClass && newClass.length > 0) {
          classIds.push(newClass[0].id);
        }
      }
      idx++;
    }

    // Link to student_classes
    const {error: delErr} = await supabase.from('student_classes').delete().eq('student_id', sId);
    if(delErr) console.error("Delete student classes err:", delErr);

    for (const cid of classIds) {
      const {error: insSC} = await supabase.from('student_classes').insert({
        student_id: sId,
        class_id: cid
      });
      if(insSC) console.error("Insert student_class err:", insSC);
    }
  }
  console.log("Done populating classes!");
}

fixClasses().catch(console.error);
